### PR TITLE
Add POST /_search/clear_scroll endpoint and deprecate delete scroll endpoint

### PR DESCRIPTION
**Pull request by javanna**
The clear scroll api currently allows to provide a scroll by specifying it either as part of the url (it is effectively the resource that gets deleted) or within the request body. The current api uses the DELETE method though, and we have decided to remove support for providing the request body with any DELETE endpoint in the future. In order to get to this for the next major version, we introduce the  new endpoint `POST /_search/clear_scroll` which replaces the current clear_scroll api and uses POST instead of DELETE. It allows to provide the `scroll_id` as a url parameter, which is though deprecated (will output a deprecation warning when used) in favour of providing it as part of the request body.

 The `DELETE /_search/scroll/` is deprecated, hence it will output a deprecation warning whenever used. The DELETE endpoints will be removed in 6.0, as well as the support for providing the scroll_id as a url parameter against the POST endpoint.

Relates to #8217
Relates to #21453

### CODE REVIEW

**Comment by polyfractal** [comment]
Technically, this break will work fine for the clients in regards to testing.  E.g. if the client doesn't support the new `delete_scroll`, it'll just fail the test (as it should).

But for the end-user, this is gonna be problematic.  If an end-user is currently using `clear_scroll`, they provide the scroll ID as part of the endpoint params, e.g:

```php
$client->clear_scroll([
  'scrollId' => 'abc123'
]);
```

But this change will make that illegal, so they have to adjust their code to move it into the body:

```php
$client->clear_scroll([
  'body' => [
    'scrollId' => 'abc123'
  ]
]);
```

Or change code to use the new `delete_scroll()` endpoint.  Either way, it's a code-change in a minor release.

So any of the clients following SemVer are going to be in a sticky place, since this is a pretty noticeable API breaking change for a 5.x release.  They'd really have to bump the major version to allow this, or start doing magic under the covers to insert the scroll ID into the body (which isn't currently done anywhere, to my knowledge).

To summarize:

- `clear_scroll` -> `delete_scroll` is fine
- Adding back a POST `clear_scroll` is problematic, since client's can't leave the old DELETE `clear_scroll` laying around as a deprecated (but usable) endpoint.  Ideally, the new POST version would be named something entirely different so it's just adding functionality, not breaking.

**Comment by javanna** [comment]
thanks a lot for your explanation @polyfractal ! I have no intentions to make a breaking change for the clients with this PR, I was just not sure it would break anything. The rename was ideal, as calling the new endpoint delete sounded funny because it uses POST. At this point I would leave clear_scroll as-is with the delete method and come up with a different name for the POST endpoint. Makes sense?

**Comment by polyfractal** [comment]
++ that'd work.  :)  

I'm afraid I don't have any ideas about an alternative name though :(

**Comment by javanna** [comment]
I made the change fully bw compatible now, @polyfractal mind having another look? Maybe @clintongormley wants to look too?

**Comment by javanna** [comment]
@elasticmachine test this please

File: rest-api-spec/src/main/resources/rest-api-spec/api/clear_scroll_post.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "clear_scroll_post": {
```
Comment (polyfractal) [inline rest-api-spec/src/main/resources/rest-api-spec/api/clear_scroll_post.json]: I wonder if this should be `post_clear_scroll`, just to follow the http-verb-preceding-noun convention (put_template, etc)?  TBH my level of caring is like 10% though :)  @elastic/es-clients any thoughts?

**Comment by polyfractal** [comment]
Left a minor comment, and I don't particularly care which way it goes.

REST stuff LGTM!  Thanks for making this client friendly :)

File: rest-api-spec/src/main/resources/rest-api-spec/api/clear_scroll_post.json (hunk 1 above)
Comment (javanna) [inline rest-api-spec/src/main/resources/rest-api-spec/api/clear_scroll_post.json]: makes sense to me

**Comment by clintongormley** [comment]
What about doing the following instead:

* `POST _search/clear_scroll` accepts a body with the scroll ID
* `POST _search/clear_scroll` also accepts the URL and query string param form of scroll_id (but add them as deprecated)
* change the `clear_scroll` REST spec to use `POST _search/clear_scroll`
* deprecate use of `DELETE _search/scroll`

This way the clients will continue to work as before, and users will get the deprecated warnings if they try to pass the scroll ID in the URL/query string.

**Comment by javanna** [comment]
@clintongormley did you mean to deprecate DELETE _search/scroll completely? That would mean that the only correct way to provide a scroll_id would be via POST and by providing it as part of the request body? I thought that we wanted to leave the DELETE _search/scroll/${scroll_id} endpoint around. Also having a new clean api that only accepts the body for POST, that could be a better trade-off than having deprecation warnings in there too? What do you think?

**Comment by clintongormley** [comment]
> did you mean to deprecate DELETE _search/scroll completely? 

Yes.  Why have two ways to do it?

> That would mean that the only correct way to provide a scroll_id would be via POST and by providing it as part of the request body?

Yes, it's the only reliable way to do it.  For the URL and QS versions, it fails when your scroll IDs are too long.

> Also having a new clean api that only accepts the body for POST, that could be a better trade-off than having deprecation warnings in there too? What do you think?

This is to handle the change in the client.  Users will keep using the clear_scroll method, but internally it'll use the `POST` endpoint.  But they may well be calling it as `clear_scroll( scroll_id => 'foo')` when they should move to passing it in the `body`.

So we don't break anything for the user in 5.x, but instead warn them that this will break in 6.0

**Comment by javanna** [comment]
thanks @clintongormley I have updated content, title and description of the PR.

File: docs/reference/search/request/scroll.asciidoc [hunk 2]
```diff
@@ -165,23 +165,24 @@ DELETE /_search/scroll
...
+All search contexts can be cleared with the special `_all` scroll id:
 
 [source,js]
 ---------------------------------------
-DELETE /_search/scroll/DXF1ZXJ5QW5kRmV0Y2gBAAAAAAAAAD4WYm9laVYtZndUQlNsdDcwakFMNjU1QQ==,DnF1ZXJ5VGhlbkZldGNoBQAAAAAAAAABFmtSWWRRWUJrU2o2ZExpSGJCVmQxYUEAAAAAAAAAAxZrUllkUVlCa1NqNmRMaUhiQlZkMWFBAAAAAAAAAAIWa1JZZFFZQmtTajZkTGlIYkJWZDFhQQAAAAAAAAAFFmtSWWRRWUJrU2o2ZExpSGJCVmQxYUEAAAAAAAAABBZrUllkUVlCa1NqNmRMaUhiQlZkMWFB
+POST /_search/clear_scroll
+{
+    "scroll_id" : [
+      "_all"
+    ]
```
Comment (javanna) [inline docs/reference/search/request/scroll.asciidoc]: this one looks a bit odd, looked better providing _all as part of the url, but maybe that's ok

Comment (clintongormley) [inline docs/reference/search/request/scroll.asciidoc]: Perhaps this could be replaced with supporting `*` as a scroll ID

File: core/src/main/java/org/elasticsearch/rest/action/search/RestClearScrollAction.java [hunk 3]
```diff
@@ -37,47 +36,61 @@
...
     public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
-        String scrollIds = request.param("scroll_id");
-        ClearScrollRequest clearRequest = new ClearScrollRequest();
-        clearRequest.setScrollIds(Arrays.asList(splitScrollIds(scrollIds)));
+        if (request.method() == DELETE) {
+            deprecationLogger.deprecated("Deprecated [DELETE /_search/scroll] endpoint used, expected [POST /_search/scroll] instead");
+        }
+        if (request.method() == POST && request.hasParam("scroll_id")) {
+            deprecationLogger.deprecated("Deprecated [POST /_search/clear_scroll/{scroll_id}] endpoint used, " +
+                    "expected [POST /_search/clear_scroll] instead. The scroll_id should rather be provided as part of the request body");
```
Comment (clintongormley) [inline core/src/main/java/org/elasticsearch/rest/action/search/RestClearScrollAction.java]: should rather -> must

File: docs/reference/search/request/scroll.asciidoc [hunk 4]
```diff
@@ -165,23 +165,24 @@ DELETE /_search/scroll
 // CONSOLE
 // TEST[catch:missing]
 
-All search contexts can be cleared with the `_all` parameter:
+The recommended way to provide the `scroll_id` is by placing it in the request
```
Comment (clintongormley) [inline docs/reference/search/request/scroll.asciidoc]: Possibly replace this block with:

    deprecated[5.2.0,Providing the scroll ID as part of the URL or as a query string parameter has been deprecated in favour of specifying it in the request body.  The `DELETE _search/scroll` end points are deprecated in favour of `POST _search/clear_scroll]

File: rest-api-spec/src/main/resources/rest-api-spec/api/clear_scroll.json [hunk 5]
```diff
@@ -14,7 +14,7 @@
       "params": {}
     },
     "body": {
-      "description": "A comma-separated list of scroll IDs to clear if none was specified via the scroll_id parameter"
+      "description": "A comma-separated list of scroll IDs to clear, has the precedence over the scroll_id parameter"
```
Comment (clintongormley) [inline rest-api-spec/src/main/resources/rest-api-spec/api/clear_scroll.json]: has precedence

File: rest-api-spec/src/main/resources/rest-api-spec/api/delete_scroll.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (clintongormley) [inline rest-api-spec/src/main/resources/rest-api-spec/api/delete_scroll.json]: Why add this at all?  This will result in adding a new method to the clients.

Comment (clintongormley) [inline rest-api-spec/src/main/resources/rest-api-spec/api/delete_scroll.json]: Is it just there for bwc testing?

File: docs/reference/search/request/scroll.asciidoc (hunk 2 above)
Comment (clintongormley) [inline docs/reference/search/request/scroll.asciidoc]: Actually we probably don't support wildcard matching, so using `_all` is better

File: rest-api-spec/src/main/resources/rest-api-spec/api/delete_scroll.json (hunk 6 above)
Comment (javanna) [inline rest-api-spec/src/main/resources/rest-api-spec/api/delete_scroll.json]: no you are right. it is a leftover, due to the fact that I thought POST should support the query_string parameter at all, now that it does, I can get right of this new api.

Comment (javanna) [inline rest-api-spec/src/main/resources/rest-api-spec/api/delete_scroll.json]: oh wait though, DELETE is done against _search/scroll while POST is done against _search/clear_scroll . I don't think we can express that in our spec. POST _search/scroll is already taken by the search api.

**Comment by javanna** [comment]
quick update: this depends on #21888 otherwise it becomes very hard to properly test the change.

**Comment by javanna** [comment]
why was this closed @martijnvg ?

**Comment by cbuescher** [comment]
@javanna is this PR still something we want to do? 
//cc @elastic/es-search-aggs

**Comment by javanna** [comment]
yes
